presto-array/src/main/java/com/facebook/presto/array/ReferenceCountMap.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.array;
+
+import io.airlift.slice.SizeOf;
+import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
+import org.openjdk.jol.info.ClassLayout;
+
+public final class ReferenceCountMap
+        extends Object2IntOpenHashMap<Object>
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(ReferenceCountMap.class).instanceSize() + ClassLayout.parseClass(Object2IntOpenHashMap.class).instanceSize();
+
+    // Two different blocks can share the same underlying data
+    // Use the visitors to avoid memory over counting
+    private final Object2IntOpenHashMap trackedObjects = new Object2IntOpenHashMap<>();
```

Comment:
I thought we were going to use a `Object2IntOpenCustomHashMap` with an identity `Strategy` so we basically get an `IdentityObject2IntMap`